File: haskell/collatz-conjecture/src/CollatzConjecture.hs
module CollatzConjecture (collatz) where

collatz :: Integer -> Maybe Integer
collatz number = if number > 0 then Just (collatz' 0 number) else Nothing
  where
    collatz' acc 1 = acc
    collatz' acc n = collatz' (acc + 1) (next n)
    next n = if even n then div n 2 else 3 * n + 1

File: solutions/elixir/collatz-conjecture/1/lib/collatz_conjecture.ex
defmodule CollatzConjecture do
  require Integer

  @doc """
  calc/1 takes an integer and returns the number of steps required to get the
  number to 1 when following the rules:
    - if number is odd, multiply with 3 and add 1
    - if number is even, divide by 2
  """
  @spec calc(input :: pos_integer()) :: non_neg_integer()
  def calc(1), do: 0
  def calc(odd) when Integer.is_odd(odd), do: 1 + calc(3 * odd + 1)
  def calc(even) when is_integer(even) and even > 0, do: 1 + calc(div(even, 2))
end
